### PR TITLE
Add firebase support to web_dashboard

### PR DESCRIPTION
**Pull request by johnpryan**

### CODE REVIEW

File: experimental/web_dashboard/tool/grind.dart [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+@Task()
+fixCopyright() async {
+  await for (var file in _filesWithoutCopyright()) {
+    var contents = await file.readAsString();
+    file.writeAsString(_copyright + '\n\n' + contents);
+  }
+}
+
+Stream<File> _filesWithoutCopyright() async* {
```
Comment (RedBrogdon) [inline experimental/web_dashboard/tool/grind.dart:86]: You should make a build_runner plugin out of this and publish it for use in the rest of our projects. 😆

File: experimental/web_dashboard/lib/src/api/api.dart [hunk 2]
```diff
@@ -2,44 +2,107 @@
...
+    return '<Category id=$id>';
+  }
 }
 
 /// A number tracked at a point in time.
+@JsonSerializable()
 class Entry {
-  final int value;
-  final DateTime time;
+  int value;
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/api/api.dart:75]: That's a good question - right now the firebase schema has entries stored in the Category document, so any backreference to the category would need to be added after parsing. There's no use-case for it yet, so I left it out.

File: experimental/web_dashboard/lib/src/api/api.dart [hunk 3]
```diff
@@ -2,44 +2,107 @@
...
-  Future<Item> delete(String id);
-  Future<Item> get(String id);
-  Future<Item> insert(Item item);
-  Future<List<Item>> list();
-  Future<Item> update(Item item, String id);
-  Stream<List<Item>> allItemsStream();
-}
+/// Manipulates [Category] data.
+abstract class CategoryApi {
+  Future<Category> delete(String id);
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/api/api.dart:18]: It's a common pattern for APIs, it's used a bit in the tests, but not in the app.

File: experimental/web_dashboard/lib/src/api/api.dart [hunk 4]
```diff
@@ -2,44 +2,107 @@
...
 
-/// Something being tracked.
-class Item {
-  final String name;
-  String id;
+  Future<Category> get(String id);
+
+  Future<Category> insert(Category category);
+
+  Future<List<Category>> list();
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/api/api.dart:24]: I think we can take out `latest` - `list` and `stream` are both required since `stream` doesn't return the initial value. (We could eventually switch these to a single RxDart stream to fix this.)

[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/api/api.dart:24]: removed `latest`

File: experimental/web_dashboard/lib/src/api/firebase.dart [hunk 5]
```diff
@@ -1,3 +1,160 @@
...
+  @override
+  Future<Entry> get(String categoryId, String id) async {
+    var document =
+        firestore.document('users/$userId/categories/$categoryId/entries/$id');
+    var snapshot = await document.get();
+    return Entry.fromJson(snapshot.data)..id = snapshot.documentID;
+  }
+}
+
+class FirebaseCategoryApi implements CategoryApi {
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/api/firebase.dart:90]: That's a good point, I added a new field for the `'users/$userId/categories'` collection reference and it improved the readability I think (f8ba724a550dba11b51dc97c930b900ce6a995ae).

File: experimental/web_dashboard/lib/src/api/mock.dart [hunk 6]
```diff
@@ -13,49 +14,77 @@ class MockDashboardApi implements DashboardApi {
...
-class MockItemApi implements ItemApi {
-  Map<String, Item> _storage = {};
-  StreamController<List<Item>> _streamController =
-      StreamController<List<Item>>.broadcast();
+class MockCategoryApi implements CategoryApi {
+  Map<String, Category> _storage = {};
+  StreamController<List<Category>> _streamController =
+      StreamController<List<Category>>.broadcast();
+
+  Stream<List<Category>> _stream;
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/api/mock.dart:44]: BroadcastStreamController returns a new stream each time the `stream` getter is called, but it looks like StreamBuilder can be given a new stream and resubscribe. My thought was that it was probably incorrect (in Angular it's a bad idea to pass a different stream to a component)

Things are still working without this field, so I took it out.

File: experimental/web_dashboard/lib/src/api/mock.dart [hunk 7]
```diff
@@ -67,40 +96,52 @@ class MockEntryApi implements EntryApi {
...
+    // TODO: limit events to those matching categoryId
+    // This stream includes events for other categories, not just the one
+    // specified by [categoryId]. To do this properly, _emit() must include the
+    // category id in all events, so that this stream can filter based off of that.
     return _streamController.stream;
   }
 
-  void _emit() {
-    _streamController.add(_storage.values.toList());
+  void _emit(String categoryId) {
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/api/mock.dart:126]: Right now it's not exactly a toggle switch, it just adds a new event to the stream with the entries for that particular category. 

Added 9280b517ba4233a55c25743bf5aa5dc522669326 to address this.

File: experimental/web_dashboard/lib/src/auth/mock.dart [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import 'auth.dart';
+
+class MockAuthService implements Auth {
+  @override
+  Future<User> signIn() async {
+    return MockUser();
+  }
+
+  @override
+  Future signOut() async {
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/auth/mock.dart:14]: I don't think we use a lint rule that catches this, but I added `analysis_options.yaml` now.

File: experimental/web_dashboard/lib/src/pages/dashboard.dart [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  @override
+  Widget build(BuildContext context) {
+    var api = Provider.of<AppState>(context).api;
+    return Scrollbar(
+      child: GridView(
+        gridDelegate: SliverGridDelegateWithMaxCrossAxisExtent(
+          childAspectRatio: 2,
+          maxCrossAxisExtent: 500,
+        ),
+        children: categories.map((category) {
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/pages/dashboard.dart:53]: much cleaner, I should take a look where we can do this elsewhere...

File: experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  Category _selected;
+  List<Category> _categories = [];
+  StreamSubscription _subscription;
+
+  void initState() {
+    super.initState();
+    _fetch();
+    _subscription = widget.api.stream().listen((categories) {
+      setState(() {
+        _categories = categories;
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart:35]: Red screen of death... It can happen if you delete a category on another device, so I added 25041dec6bafa28746dfa4c034ded0249214dd93 to handle this case.

File: experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+import 'package:web_dashboard/src/api/api.dart';
+
+/// Subscribes to the latest list of categories and allows the user to select
+/// one.
+class CategoryDropdown extends StatefulWidget {
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart:12]: Moving the subscription stuff means that the parent widget has to manage that, and since this widget is used in a couple of places, I'd like to keep it here.

File: experimental/web_dashboard/lib/src/pages/entries.dart [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+                    ),
+                    FlatButton(
+                      child: Text('Delete'),
+                      onPressed: () => Navigator.of(context).pop(true),
+                    ),
+                  ],
+                ),
+              );
+              if (shouldDelete) {
+                Provider.of<AppState>(context, listen: false)
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/pages/entries.dart]: yep, fixed.

**Comment by johnpryan** [comment]
@RedBrogdon b677d2e works around the [firebase issue](https://github.com/FirebaseExtended/flutterfire/issues/2508) I ran into, so this should be ready for another look.

File: experimental/web_dashboard/tool/grind.dart (hunk 1 above)
Comment (johnpryan) [reply experimental/web_dashboard/tool/grind.dart:88]: Sounds good, I think I'll follow up with that change in a separate PR

File: experimental/web_dashboard/README.md [hunk 13]
```diff
@@ -30,3 +30,107 @@ Skia / CanvasKit mode:
...
+
+## Add Firebase
+
+### Step 1: Create Firebase project and install packages
+
+Follow the steps at [firebase.google.com/docs/flutter/setup][flutter-setup].
+This app uses these Firebase packages:
+
+```yaml
+# Add these packages to your pubspec.yaml under "dependencies"
```
Comment (RedBrogdon) [inline experimental/web_dashboard/README.md:47]: These are already in the pubspec, aren't they?

File: experimental/web_dashboard/README.md [hunk 14]
```diff
@@ -30,3 +30,107 @@ Skia / CanvasKit mode:
...
+
+### Step 6: Run the app
+
+Run the app on port 5000:
+
+```bash
+flutter run -d chrome --web-port=5000
+```
+
+If you see CORS issues, go to the [Services section][cloud-console-apis] in the
```
Comment (RedBrogdon) [inline experimental/web_dashboard/README.md:124]: Consider adding "CORS issues in your browser's console" or something similar, just to point new devs in the right direction.

File: experimental/web_dashboard/lib/src/api/firebase.dart [hunk 15]
```diff
@@ -1,3 +1,160 @@
...
+    });
+
+    return result;
+  }
+
+  @override
+  Future<Entry> delete(String categoryId, String id) async {
+    var document =
+        firestore.document('users/$userId/categories/$categoryId/entries/$id');
+    var category = await get(categoryId, document.documentID);
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply experimental/web_dashboard/lib/src/api/firebase.dart:45]: Gentle ping here. It looks like this variable is still misnamed.

File: experimental/web_dashboard/lib/src/app.dart [hunk 16]
```diff
@@ -2,58 +2,102 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+import 'package:cloud_firestore/cloud_firestore.dart';
 import 'package:flutter/material.dart';
 import 'package:provider/provider.dart';
+import 'package:web_dashboard/src/api/api.dart';
```
Comment (RedBrogdon) [inline experimental/web_dashboard/lib/src/app.dart:8]: Were these auto-imported by IntelliJ using the wrong schema? It looks like the old imports are path-based, and these new ones are package-based.

File: experimental/web_dashboard/lib/src/pages/dashboard.dart [hunk 17]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2020, the Flutter project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:provider/provider.dart';
+import 'package:web_dashboard/src/api/api.dart';
```
Comment (RedBrogdon) [inline experimental/web_dashboard/lib/src/pages/dashboard.dart:7]: This is referenced in an earlier comment as well, but it looks like some imports of internal files are `package:`-style and some are direct paths. Whichever one you prefer is fine, but I'd pick a type and stick with it for readability.

File: experimental/web_dashboard/lib/src/pages/entries.dart [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+  EntryTile({
+    this.category,
+    this.entry,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    return ListTile(
+      title: Text(entry.value.toString()),
```
Comment (RedBrogdon) [inline experimental/web_dashboard/lib/src/pages/entries.dart:110]: nit: I've been told by folks on the Dart team that string interpolation is preferred to `toString`:

```
      title: Text('${entry.value}');
```

File: experimental/web_dashboard/lib/src/utils/chart_utils.dart [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        es.add(entry);
+      }
+    }
+
+    yield es;
+    d = d.add(Duration(days: 1));
+  }
+}
+
+/// The UTC date of a datetime, excluding the minutes, seconds, etc.
```
Comment (RedBrogdon) [inline experimental/web_dashboard/lib/src/utils/chart_utils.dart:62]: Could you use extension methods for these last two?

File: experimental/web_dashboard/lib/src/widgets/forms.dart [hunk 20]
```diff
@@ -0,0 +1,4 @@
+// Copyright 2020, the Flutter project authors. Please see the AUTHORS file
```
Comment (RedBrogdon) [inline experimental/web_dashboard/lib/src/widgets/forms.dart:1]: Is this file necessary?

File: experimental/web_dashboard/README.md [hunk 21]
```diff
@@ -30,3 +30,107 @@ Skia / CanvasKit mode:
...
+```
+
+### Step 2: Enable Google Sign In for your project
+
+In the Firebase console, go to "Authentication" and enable Google sign in. Click
+on "Web SDK Configuration" and copy down your Web client ID.
+
+### Step 3: Add firebase scripts to `index.html`
+
+Add these scripts to the `<head>` tag:
```
Comment (RedBrogdon) [inline experimental/web_dashboard/README.md:61]: These files are already linked into index.html in this project, right?

I think I may be misinterpreting the purpose of these instructions. Are these meant as general guidelines for getting Firebase up and running in a Flutter app built for web, or for this app specifically? If it's the former, I'm not sure that content fits this location. If it's the latter, I'd recommend cutting anything already done in the project.

File: experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart (hunk 11 above)
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart:12]: This is the only part of the code I'm not completely clear on. Let's make some time in an upcoming 1:1 to walk through it.

The reason I'm balking is that there's a decent bit of code dedicated to managing the stream subscription and calling `setState`, which is StreamBuilder's purpose in life. That's making me wonder if you can simplify things here by finding a way to use StreamBuilder. To be clear, there may not be, but if there is a way to do it, it's probably a good idea.

**Comment by RedBrogdon** [comment]
Added a bunch of nits and one real question. This is definitely in the home stretch.

File: experimental/web_dashboard/lib/src/pages/entries.dart (hunk 18 above)
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/pages/entries.dart:108]: You're right, I think it's [in the style guide](https://dart.dev/guides/language/effective-dart/usage#prefer-using-interpolation-to-compose-strings-and-values)

File: experimental/web_dashboard/lib/src/utils/chart_utils.dart (hunk 19 above)
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/utils/chart_utils.dart]: Yeah it's a little cleaner: https://github.com/flutter/samples/pull/421/commits/3b94049258505fddd91f8aa643fbd27d49e108d9

File: experimental/web_dashboard/README.md (hunk 21 above)
Comment (johnpryan) [reply experimental/web_dashboard/README.md]: Decided go with option 2. and made the instructions specific for this sample 7e7df2bf91da613e40d67c9502d77be2d8d7fd03

File: experimental/web_dashboard/lib/src/widgets/forms.dart (hunk 20 above)
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/widgets/forms.dart]: oops, good catch.

File: experimental/web_dashboard/lib/src/pages/dashboard.dart (hunk 17 above)
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/pages/dashboard.dart]: Changed the imports to relative ones: https://github.com/flutter/samples/pull/421/commits/4e47b92109adce3efaf29cff826ee254327e2dd9

File: experimental/web_dashboard/lib/src/utils/day_helpers.dart [hunk 22]
```diff
@@ -0,0 +1,11 @@
+extension DayUtils on DateTime {
```
Comment (RedBrogdon) [inline experimental/web_dashboard/lib/src/utils/day_helpers.dart:1]: nit: copyright header.

Comment (RedBrogdon) [reply experimental/web_dashboard/lib/src/utils/day_helpers.dart:1]: I'm chuffed to see a good use case for extension methods in one of our samples. 😄

File: experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart [hunk 23]
```diff
@@ -24,61 +24,67 @@ class CategoryDropdown extends StatefulWidget {
...
-    _fetch();
-    _subscription = widget.api.subscribe().listen((categories) {
-      setState(() {
-        _categories = categories;
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder<List<Category>>(
+      future: widget.api.list(),
+      builder: (context, futureSnapshot) {
+        // Show an empty dropdown while the data is loading
```
Comment (RedBrogdon) [inline experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart:33]: nit: period at the end of the sentence.

File: experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart [hunk 24]
```diff
@@ -24,61 +24,67 @@ class CategoryDropdown extends StatefulWidget {
...
     }
+    _selected = category;
 
-    setState(() {
-      _categories = categories;
+    // Avoid invoking this callback during a build() because 1. calling setState
+    // during a build is an error, and 2. any widgets listening could be calling
+    // setState(). This can happen if the data is immediately returned by the
+    // FutureBuilder or StreamBuilder.
+    Future(() {
```
Comment (RedBrogdon) [inline experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart:81]: This is an interesting problem. The app needs the list of categories to set the value of _selected, but it also needs the Future that's providing them to use in the FutureBuilder.

One thing you might do is try to move that assignment into a different place. I believe you could:

* Add an override for initState that
  - invokes widget.api.list()
  - stores the Future in a field in the State object
  - uses `then` to add a clause calling setState, assigning _selected, and returning the same list of categories.
* That stored Future could then be given to the FutureBuilder.
* The builder methods for the FutureBuilder/StreamBuilder could check the value of _selected, and (if it's null) just display an empty list as they do now if there's no data yet.

It's also possible that you could pull this off with a single call to `subscribe` (assuming it's a broadcast stream), by getting a reference to the stream in initState, assigning the stream to a property in the state object for use in `build`, and then using `.first.then` on the stream to call setState once you've gotten at least one value.

This is all speculative since I haven't tried it (and probably too large a suggestion), so I'm just throwing it out here while also LGTMing the PR. 😄

File: experimental/web_dashboard/lib/src/utils/day_helpers.dart (hunk 22 above)
Comment (RedBrogdon) [reply experimental/web_dashboard/lib/src/utils/day_helpers.dart:1]: Also, I've been watching so much of the Great British Bake-Off during quarantine that apparently I now say "chuffed."

File: experimental/web_dashboard/android/app/build.gradle [hunk 25]
```diff
@@ -39,7 +39,7 @@ android {
     defaultConfig {
         // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
         applicationId "dev.flutter.web_dashboard"
-        minSdkVersion 16
+        minSdkVersion 21
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply experimental/web_dashboard/android/app/build.gradle:42]: Yeah I was seeing this exception: "The number of method references in a .dex
file cannot exceed 64K."

I think I found this solution here: https://stackoverflow.com/a/38932169/5859156

File: experimental/web_dashboard/lib/src/utils/day_helpers.dart (hunk 22 above)
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/utils/day_helpers.dart:1]: good bake, err, catch.

File: experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart (hunk 24 above)
Comment (johnpryan) [reply experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart]: This is a good suggestion, here's how this ended up: https://github.com/flutter/samples/pull/421/commits/34796cec49b54f8cfd460e71ac7dc1c04cbf4a3e
